### ISSUE TITLE
BEM class naming?

### ISSUE THREAD

**Issue by PixelT** [uikit/uikit#1123]
As @saschadube say here:
https://github.com/uikit/uikit/issues/800#issuecomment-67820754
In UIkit 3 the class will be renamed - maybe this is good time to consider to using the BEM methodology in class names?

**Comment by zslabs** [comment]
I had brought this up in Gitter, but I'm not sure if they've decided on this as of yet.

**Comment by florianletsch** [comment]
I'm torn on this. The aundance of `-`, `--`, `__`, `_` just looks... weird to me. This is just a personal perception of course, the current names with just single dashes feel cleaner. 

I still get it though, right now there's no distinction between modifier classes and sub components (aka *elements* according to BEM), and strict BEM would help with that.

What is your personal take on this? Did you have the same doubts in the beginning but just followed along and got used to what the classes look like? Or am I just being picky here? :)

What does BEM say about state classes (`uk-active`) and utility classes (`uk-text-center`) anyway, how would those work out?

**Comment by krzksz** [comment]
I am a strong +1 for this. I have been already combining my BEM CSS with uikit's one using brackets to visually group classes and this would greatly unify my style e.x.:
```html
<form class="login-form [ uk-form ]">
    <input type="text" placeholder="" class="login-form__input [ uk-form-large ]">
    <input type="text" placeholder="" class="login-form__input [ uk-form-small ]">
</form>```

@florianletsch As to utility classes, we could simply have `uk-text--center`, `uk-float--left`, `uk-float--right` etc. which would make sense to me. As to `uk-active` inside `nav` component it could become `uk-nav__item--active`:

```html
<ul class="uk-nav uk-nav--parent-icon">
    <li class="uk-nav__item uk-nav__item--parent">
        <a href="#">...</a>
        <ul class="uk-nav--sub">
            <li class=uk-nav__item uk-nav__item--active>
                <a href="#">...</a>
            </li>
        </ul>
    </li>
</ul>```

What do you guys think?

**Comment by zslabs** [comment]
@florianletsch Having a little trouble finding my thoughts on the subject I posted here, but Chris Coyier actually just wrote a good article on the subject today https://css-tricks.com/bem-101/

I completely get the dashes are underscores are a bit odd to look at, but I went into BEM with an open mind after thinking it'd be difficult to use for so long, but have used it on every project so far (including one that scales to over 50 million visitors a month, with a humungous codebase). It's kept things lean, mean and focused. Think the core of UIkit being written in a BEM style would be a fantastic compliment.

@krzksz gave a great example there too.
